awscli/customizations/s3events.py

```diff
@@ -57,6 +65,46 @@ def replace_event_stream_docs(help_command, **kwargs):
               "object content is written to the specified outfile.\n")
 
 
+def document_expires_string(help_command, **kwargs):
```

Comment:
There doesn't seem to be a cleaner way to handle this with the existing interfaces. [Writes are just stored in a list](https://github.com/aws/aws-cli/blob/c6305d03c636481d39016b678ed9b3405fc3f0b7/awscli/botocore/docs/bcdoc/restdoc.py#L32) in `ReSTDocument` objects, so I wonder if it's worth implementing a couple extra methods to help simplify injecting lines.
```
# Only copying/including relevant details
class ReSTDocument(object):
    def __init__(self, target='man'):
        self._writes = []

    def find_write(self, content):
        try:
            return self._writes.index(content)
        except ValueError:
            return None

    def insert_write(self, before_index, content):
        # This may need some massaging for correct formatting
        self._writes.insert(before_index, content)
```

Then in `document_expires_string` we could return early if `doc.find_write('Expires -> (timestamp)'` returns `None`. Otherwise we inject writes.

Thoughts?

Reply:
Also, in the past, when we had to [add notes for specific types](https://github.com/aws/aws-cli/blob/c6305d03c636481d39016b678ed9b3405fc3f0b7/awscli/clidocs.py#L291-L308), we implemented them in `CLIDocumentEventHandler` so the notes would get written at the same time as the rest of the docs and not after. I'm not sure if we want to include S3-specific logic here but it's another option.

Reply:
In reply to the suggestions for find and insert writes, I prefer your suggestions and will implement them in the next revision.

In terms of CLIDocumentEventHandler, I think it'd make more sense to have it under the customizations directory since it's service-specific. But I think a generic `inject_note_on_output_field(service_name, field_name, field_type, note_description)` function and `inject_output_field_after(service_name, prior_field_name, prior_field_type, synthetic_field_name, synthetic_field_type, synthetic_field_desc)` function could be useful, and we could call it from s3events to make that code a lot cleaner and simpler.

The first function is self-documenting in its name, but the second function I'd imagine would inject a new field named `synthetic_field_name` with type `synthetic_field_type` with description `synthetic_field_desc`. Specifically, it would inject this field right after the field with name `prior_field_name` and type `prior_field_type` if such field exists under service `service_name`. If no such field is found under such service, the synthetic field will not be injected. 

But maybe this is too specific and we should wait for at least a couple duplications of this use-case before abstracting it in this way?

Reply:
> But maybe this is too specific and we should wait for at least a couple duplications of this use-case before abstracting it in this way?

Yeah let's go with a simpler solution for now and we can revisit it if we find ourselves doing this more often.